proposals/NNNN-clock-instant-date-duration.md

```diff
@@ -0,0 +1,498 @@
+
+```swift
+public struct WallClock {
+  public init()
+  
+  public static var now: Date { get }
+}
+
+extension WallClock: ClockProtocol {
+  public typealias Instant = Date
+
+  public var now: Date { get }
+  public func sleep(until deadline: Date) async throws
+}
+
+extension ClockProtocol where Self == WallClock {
+  public static var wall: WallClock { get }
+}
+```
+
+##### MonotonicClock
+
+When instants are for local processing only and need to be high resolution without the encumbrance of suspension while the machine is asleep `MonotonicClock` is the tool for the job. The `MonotonicClock.Instant` type can be initialized with a wall clock instant if that value can be expressed in terms of a relative point to now; knowing the delta between the current time and the specified wall clock instant a conversion to the current monotonic reference point can be made such that conversion (if possible) represents what the value would be in terms of the monotonic clock. Much like the wall clock version the monotonic clock also offers an extension to access the clock instance as the inferred base type property.
+
+```swift
+public struct MonotonicClock {
+  public init()
+  
+  public static var now: Instant { get }
+}
+
+extension MonotonicClock: ClockProtocol {
+  public struct Instant { 
+    public init?(converting wallclockInstant: WallClock.Instant)
+    
+    public static var now: MonotonicClock.Instant { get }
+  }
+
+  public var now: Instant { get }
+  public func sleep(until deadline: Instant) async throws
+}
+
+extension MonotonicClock.Instant: InstantProtocol { 
+  func advanced(by duration: Duration) -> MonotonicClock.Instant
+  func duration(to other: MonotonicClock.Instant) -> Duration
+}
+
+extension ClockProtocol where Self == MonotonicClock {
+  public static var monotonic: MonotonicClock { get }
+}
+```
+
+##### UptimeClock
+
+Where local process scoped or cross machine scoped instants are not suitable uptime serves the purpose of a clock that does not increment while the machine is asleep but is a time that is referenced to the boot time of the machine, this allows for the affordance of cross process communication in the scope of that machine. Similar to the other clocks there is an extension to access the clock instance as the inferred base type property.
+
+```swift
+public struct UptimeClock: ClockProtocol {
+  public init()
+  
+  public static var now: Instant { get }
+}
+
+extension UptimeClock: ClockProtocol {
+  public struct Instant { 
+    public init?(converting wallclockInstant: WallClock.Instant)
+    public static var now: UptimeClock.Instant { get }
```

Comment:
Same comment here.

Reply:
Same answer (Technically Date has that too and has super handy for folks)

Reply:
Right. I guess I was accustomed to `Date.now` from the latest Foundation version.